src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,138 @@
+    private int size;
+
+    public MyLinkedList() {
+        head = new Node<>(null, null, null);
+        tail = head;
+        size = 0;
+    }
+
+    private void increaseSize(int value) {
+        size += value;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("The specified index is outside the list");
+        }
+    }
+
+    private Node<T> searchNode(int index) {
+        checkIndex(index);
+        Node<T> currentNode = head;
+
+        int currentIndex = 0;
+        while (index != currentIndex) {
+            currentNode = currentNode.next;
+            currentIndex++;
+        }
+        return currentNode;
+    }
+
     @Override
     public void add(T value) {
+        if (isEmpty()) {
+            head.value = value;
+            tail = head;
+        } else {
+            tail.next = new Node<>(value, null, tail);
+            tail = tail.next;
+        }
+        increaseSize(1);
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else if (index == 0) {
+            head = new Node<>(value, head, null);
+            increaseSize(1);
+        } else {
+            Node<T> currentNode = searchNode(index);
+            currentNode.prev.next = new Node<>(value, currentNode, currentNode.prev);
+            currentNode.prev = currentNode.prev.next;
+            increaseSize(1);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return searchNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> currentNode = searchNode(index);
+        T oldNodeValue = currentNode.value;
+        currentNode.value = value;
+        return oldNodeValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> currentNode = searchNode(index);
+        if (index == (size - 1)) {
+            tail = tail.prev;
+            if (head.next != null) {
+                tail.next = null;
+            }
+        } else if (index == 0) {
+            head.next.prev = null;
+            head = head.next;
+        } else {
+            currentNode.next.prev = currentNode.prev;
+            currentNode.prev.next = currentNode.next;
+        }
+        increaseSize(-1);
+        return currentNode.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> currentNode = head;
+        for (int index = 0; index < size(); index++) {
+            if (currentNode.value == object
+                    || object != null && object.equals(currentNode.value)) {
+                remove(index);
+                return true;
+            }
+            currentNode = currentNode.next;
+        }
```

Comment:
with current implementation `remove(T object)` you iterates over list twice.
First, when you get the index.
Second, in the method `remove(int index)`, when you get the node. 

you can simplify it by adding `unlink(Node<T> node)` method.

Create a separate method`unlink(Node node)`.

See common mistakes https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html